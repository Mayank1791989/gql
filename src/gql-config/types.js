/* @flow */
/* @babel-flow-runtime-enable */
/* eslint-disable no-use-before-define */
import {
  type IQueryParser,
  type ISchemaParser,
  type IValidationContext,
  type SchemaPluginParams,
  type ISchemaPlugin,
} from 'gql-shared/types';
import { type FragmentScope } from 'gql-shared/FragmentScope';
import {
  type ValidateConfig,
  type ValidateConfigResolved,
} from 'gql-shared/GQLValidate/types';
import { type ASTVisitor } from 'graphql';
import { reify, type Type } from 'flow-runtime';
import semver from 'semver';

export type Globs = string | Array<string>;
export type FileMatchConfig = Globs | { include: Globs, ignore?: Globs };
type IQueryContext = any;

export type SemverVersion = string;
(reify: Type<SemverVersion>).addConstraint((input: string) => {
  try {
    // eslint-disable-next-line no-new
    new semver.Range(input);
  } catch (err) {
    return 'Not a valid semver version.';
  }

  return null;
});

type SchemaValidationRule = { +create: (context: $FixMe) => ASTVisitor };
type QueryValidationRule = { +create: (context: $FixMe) => ASTVisitor };

export type GQLConfigFile = {|
  schema: SchemaConfig,
  query?: {
    files: Array<QueryConfig>,
  },
  version?: SemverVersion,
  options?: OptionsConfig,
|};

export type SchemaConfig = {|
  files: FileMatchConfig,
  parser?: PkgConfig,
  validate?: ValidateConfig<SchemaValidationRule>,
  presets?: Array<PkgConfig>,
  graphQLOptions?: GraphQLOptions,
  plugins: $ReadOnlyArray<PkgConfig>,
|};

export type QueryConfig = {|
  match: FileMatchConfig,
  parser?: PkgConfig,
  validate?: ValidateConfig<QueryValidationRule>,
  presets?: Array<PkgConfig>,
  fragmentScopes?: Array<FragmentScope>,
|};

export type OptionsConfig = {|
  modulePaths?: $ReadOnlyArray<string>,
|};

export type GQLConfigFileResolved = {|
  schema: SchemaConfigResolved,
  query?: {
    files: $ReadOnlyArray<QueryConfigResolved>,
  },
  version: ?SemverVersion,
  options: OptionsConfigResolved,
|};

export type SchemaConfigResolved = {|
  files: FileMatchConfig,
  validate: ValidateConfigResolved<SchemaValidationRule>,
  parser: SchemaParserLoadedPkg,
  extendSchema: Array<{ presetName: string, getSchema: () => string }>,
  graphQLOptions?: GraphQLOptions,
  plugins: $ReadOnlyArray<SchemaPluginLoadedPkg>,
|};

export type QueryConfigResolved = {|
  match: FileMatchConfig,
  parser: QueryParserLoadedPkg,
  validate: ValidateConfigResolved<QueryValidationRule>,
  extendSchema: Array<{ presetName: string, getSchema: () => string }>,
  fragmentScopes: Array<FragmentScope>,
  QueryContext?: ?IQueryContext,
|};

export type OptionsConfigResolved = {|
  modulePaths: $ReadOnlyArray<string>,
|};

export type GraphQLOptions = {|
  commentDescriptions?: boolean,
  allowLegacySDLEmptyFields?: boolean,
  allowLegacySDLImplementsInterfaces?: boolean,
|};

type PkgOptions = Object;
type PkgName = string;
export type PkgConfig = PkgName | [PkgName, PkgOptions];

export type SchemaParserLoadedPkg = {
  create(): ISchemaParser,
};
export type QueryParserLoadedPkg = {
  create(): IQueryParser,
};
export type SchemaPluginLoadedPkg = {
  create(params: SchemaPluginParams): ISchemaPlugin,
};

export type QueryPreset = {
  parser?: ?PkgConfig,
  parserOptions?: Object,
  extendSchema?: ?() => string,
  validate?: ?ValidateConfig<QueryValidationRule>,
  fragmentScopes?: Array<FragmentScope>,
  __overrides?: {
    // TODO: validate
    ValidationContext?: IValidationContext,
    QueryContext?: IQueryContext,
  },
};

export type QueryPresetMerged = {|
  parser: PkgConfig,
  parserOptions: Object,
  extendSchema: Array<{ presetName: string, getSchema: () => string }>,
  validate: ValidateConfig<QueryValidationRule>,
  fragmentScopes?: Array<FragmentScope>,
  __overrides: {
    ValidationContext?: IValidationContext,
    QueryContext?: IQueryContext,
  },
|};

export type SchemaPreset = {
  parser?: PkgConfig,
  parserOptions?: Object,
  extendSchema?: ?() => string,
  validate?: ?ValidateConfig<SchemaValidationRule>,
};

export type SchemaPresetMerged = {|
  parser: PkgConfig,
  parserOptions: Object,
  extendSchema: Array<{ presetName: string, getSchema: () => string }>,
  validate: ValidateConfig<SchemaValidationRule>,
|};

export function validateConfigFile(fileData: any): GQLConfigFile {
  const GQLConfigFileType = (reify: Type<GQLConfigFile>);
  GQLConfigFileType.assert(fileData);
  return fileData;
}
