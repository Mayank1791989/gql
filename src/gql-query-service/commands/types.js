/* @flow */
import QueryContext from '../shared/QueryContext';
import { type GQLPosition } from 'gql-shared/types';
import { Source } from 'graphql';

export type CommandParams = $ReadOnly<{|
  context: QueryContext,
  source: Source,
  position: GQLPosition,
|}>;
