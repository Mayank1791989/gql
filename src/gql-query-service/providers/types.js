/* @flow */
import QueryContext from '../shared/QueryContext';
import { type GQLPosition } from 'gql-shared/types';
import { Source } from 'graphql';
import QueryTokenTypeInfe from '../shared/QueryTokenTypeInfo';
import { ParsedQueryDocument } from 'gql-shared/GQLQueryFile';

export type ProviderParams = $ReadOnly<{|
  token: $FixMe,
  typeInfo: QueryTokenTypeInfe,
  context: QueryContext,
  source: Source,
  position: GQLPosition,
|}>;

export type RulesProviderParams = {
  context: QueryContext,
  queryDocument: ParsedQueryDocument,
};
