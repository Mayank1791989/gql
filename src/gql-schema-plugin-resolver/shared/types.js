/* @flow */
/* @babel-flow-runtime-enable */
import {
  type ValidateConfig,
  type ValidateConfigResolved,
} from 'gql-shared/GQLValidate/types';
import { type IResolverParser } from './language/parser';
import {
  type TypeSchemaResolverValidationRule,
  type TypeLocalResolverValidationRule,
  type TypeGlobalResolverValidationRule,
  type ResolverValidationRule,
} from './validation/types';
import { type FileMatchConfig, type PkgConfig } from 'gql-config/types';

export type ResolverValidationConfig = ValidateConfig<ResolverValidationRule>;

export type ResolverValidateConfigResolved = {|
  schema: ValidateConfigResolved<TypeSchemaResolverValidationRule>,
  local: ValidateConfigResolved<TypeLocalResolverValidationRule>,
  global: ValidateConfigResolved<TypeGlobalResolverValidationRule>,
|};

export type ResolverParserLoadedPkg = {
  create(): IResolverParser,
};

export type ResolverPluginOptions = {|
  files: FileMatchConfig,
  parser: PkgConfig,
  validate?: ResolverValidationConfig,
|};

export type ResolverPluginOptionsResolved = {|
  files: FileMatchConfig,
  parser: ResolverParserLoadedPkg,
  validate: ResolverValidateConfigResolved,
|};
