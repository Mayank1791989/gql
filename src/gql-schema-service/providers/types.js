/* @flow */
import { type GQLPosition } from 'gql-shared/types';
import { Source } from 'graphql';
import SchemaContext from '../shared/SchemaContext';

export type ProviderParams = {|
  token: $FixMe,
  context: SchemaContext,
  source: Source,
  position: GQLPosition,
|};

export type RulesProviderParams = {|
  context: SchemaContext,
|};
