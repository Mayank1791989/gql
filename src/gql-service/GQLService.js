/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log from 'gql-shared/log';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';
import getResolverLocation from 'gql-schema-service/commands/getResolverLocation';
import getSchemaLocation from 'gql-schema-service/commands/getSchemaLocation';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

type GetResolverOrSchemaParams = $ReadOnly<{|
  fieldType: string,
  fieldName: string,
|}>;

export type { Options as GQLServiceOptions };

export function isGql(sourcePath: string) {
  return !(sourcePath.endsWith('ts') || sourcePath.endsWith('ts'));
}

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService: this._schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  onLog(listener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      logger.time('autocomplete response');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getHintsAtPosition');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      logger.timeEnd('getHintsAtPosition');
      logger.timeEnd('autocomplete response');
      return results;
    }, []);
  }

  getLenses(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getLenses request');
      logger.time('getLenses response');
      if (!this._isRunning) {
        return null;
      }

      let defLenses = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getLenses');

        if (fileConfig && fileConfig.type === 'schema') {
          defLenses = await this._schemaService.getLenses({
            filePath: params.sourcePath,
          });
        }
      } else {
        logger.time('getLenses');
        defLenses = await this._schemaService.getLenses({
          filePath: params.sourcePath,
        });
      }

      logger.timeEnd('getLenses');
      logger.timeEnd('getLenses response');
      return defLenses;
    }, null);
  }

  getResolverLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getResolverLocation request');
      logger.time('getResolverLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getResolverLocation');

      const schema = this._schemaService.getSchema();
      const resolverLocation = await getResolverLocation(
        params.fieldType,
        params.fieldName,
        schema,
        this._config.getResolverConfig().files,
        this._config.getResolverConfig().baseDir,
      );

      logger.timeEnd('getResolverLocation');
      logger.timeEnd('getResolverLocation response');
      return resolverLocation;
    }, null);
  }

  getSchemaLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      logger.debug('getSchemaLocation request');
      logger.time('getSchemaLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getSchemaLocation');

      const schema = this._schemaService.getSchema();
      const schemaLocation = getSchemaLocation(
        params.fieldType,
        params.fieldName,
        schema,
      );

      logger.timeEnd('getSchemaLocation');
      logger.timeEnd('getSchemaLocation response');
      return schemaLocation;
    }, null);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getDef request');
      logger.time('getDef response');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getDef');
        if (fileConfig && fileConfig.type === 'schema') {
          defLocation = await this._schemaService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
            resolversGlob: this._config.getResolverConfig().files,
            resolversBaseDir: this._config.getResolverConfig().baseDir,
          });
        }

        if (this._queryService && fileConfig && fileConfig.type === 'query') {
          defLocation = await this._queryService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
          });
        }
      } else {
        logger.time('getDef');
        defLocation = await this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getDef');
      logger.timeEnd('getDef response');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      logger.time('findRefs response');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      // if (this._queryService && fileConfig && fileConfig.type === 'query') {
      //   refLocations = this._queryService.findRefsOfTokenAtPosition({
      //     fileContent: params.sourceText,
      //     filePath: params.sourcePath,
      //     fileOptions: fileConfig.opts,
      //     position: params.position,
      //   });
      // }
      logger.timeEnd('findRefs');
      logger.timeEnd('findRefs response');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      logger.time('getInfo response');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getInfoAtToken');
      logger.timeEnd('getInfo response');
      return info;
    }, null);
  }
}
