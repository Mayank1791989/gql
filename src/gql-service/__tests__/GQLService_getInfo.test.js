/* @flow */
import code from 'gql-test-utils/code';
import path from 'path';
import { getSchemaFiles, getAllInfo } from 'gql-test-utils/test-data';
import { createTempFiles } from 'gql-test-utils/file';
import { type GQLPosition } from 'gql-shared/types';
import GQLService from '../GQLService';

describe('Schema: getDef', () => {
  it('works in schema files', async () => {
    const dir = createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: Viewer
        }

        type Viewer {
          name: string
        }
      `,
    });

    const gql = new GQLService({
      configDir: dir,
      watch: false,
    });

    gql.onError(err => {
      throw err;
    });

    await gql.start();

    expect(
      gql.getInfo({
        sourcePath: path.join(dir, 'schema/user.gql'),
        ...code(`
          type User {
            viewer: Viewer
            #---------^
          }
        `),
      }),
    ).toMatchSnapshot();
  });

  it('should not throw if called before onInit', () => {
    const dir = createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: Viewer
        }

        type Viewer {
          name: string
        }
      `,
    });

    const gql = new GQLService({
      configDir: dir,
      watch: false,
    });

    gql.onError(err => {
      throw err;
    });

    const run = () =>
      gql.getInfo({
        sourcePath: path.join(dir, 'schema/user.gql'),
        ...code(`
          type User {
            viewer: Viewer
            #---------^
          }
        `),
      });

    expect(run).not.toThrow();
    expect(run()).toBe(null);
  });
});

const allInfo = getAllInfo();

describe('Schema', () => {
  test('field type: ObjectType', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test {
          field: Player,
          #--------^
        }
      `),
    });
    expect(info).toEqual(allInfo.PlayerType);
  });

  test('field type: Enum', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test {
          field: Role,
            #-----^
        }
      `),
    });
    expect(info).toEqual(allInfo.RoleEnum);
  });

  test('field type: CustomScalar', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test {
          field: CustomScalar,
            #--------^
        }
      `),
    });
    expect(info).toEqual(allInfo.CustomScalar);
  });

  test('union type', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        union Test = Player | NewPlayer;
          #------------^
      `),
    });
    expect(info).toEqual(allInfo.PlayerType);
  });

  test('arguments', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test {
          test(a: CustomScalar): string
            #--------^
        }
      `),
    });
    expect(info).toEqual(allInfo.CustomScalar);
  });

  test('implements', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test implements Node {
                      #-------^
          test: string
        }
      `),
    });
    expect(info).toEqual(allInfo.NodeInterface);
  });

  test('core types', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test {
          name: String
          #-------^
        }
      `),
    });
    expect(info).toEqual(allInfo.String);
  });

  test('unknown type', async () => {
    const info = await getInfo({
      sourcePath: 'schema/test.graphql',
      ...code(`
        type Test implements Node {
          test: xString
          #-------^
        }
      `),
    });
    expect(info).toEqual(null);
  });
});

describe('Query', () => {
  describe('fragments', () => {
    it('on Type', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
                  #-----------^
              id
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.PlayerType);
    });

    it('field: Should Include description, fieldDefn and output Type', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on NewPlayer {
              friend
              #--^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.NewPlayerFriendField);
    });

    it('field arguments', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image(size
                #----^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.PlayerImageSizeArg);
    });
  });

  describe('mutations', () => {
    it('type: include description', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            mutation {
            #--^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.Mutation);
    });

    it('field: Include both input and output type', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            mutation {
              PlayerCreate
              #-----^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.MutationPlayerCreate);
    });

    it('input object fields', async () => {
      const info = await getInfo({
        sourcePath: 'query/mutation.gql',
        ...code(`
          mutation {
            PlayerCreate(
              input: {
                id: "some_id",
                role:
            #----^
              }
            ) {
              player { id }
            }
          }
        `),
      });
      expect(info).toEqual(allInfo.PlayerCreateInputRoleField);
    });

    it('variable definition type', async () => {
      const info = await getInfo({
        sourcePath: 'query/test.graphql',
        ...code(`
          mutation PlayerCreate($input: PlayerCreateInput)
                          #----------------^
        `),
      });
      expect(info).toEqual(allInfo.PlayerCreateInput);
    });
  });

  describe('query', () => {
    it('query keyword', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            query Viewer { viewer }
            #--^
          \`
        `),
      });
      expect(info).toEqual(allInfo.Query);
    });

    it('field', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            query Viewer { viewer }
                  #-----------^
          \`
        `),
      });
      expect(info).toEqual(allInfo.QueryViewerField);
    });
  });

  describe('subscriptions', () => {
    it('type: include description', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            subscription {
          #-----^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.Subscription);
    });

    it('field: Include both input and output type', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            subscription {
              LikeStory
              #-----^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.SubscriptionLikeStoryField);
    });

    it('input object fields', async () => {
      const info = await getInfo({
        sourcePath: 'query/test.graphql',
        ...code(`
          subscription {
            LikeStory(
              input: {
                id: ,
            #----^
              }
            ) {
              clientSubscriptionId
            }
          }
        `),
      });
      expect(info).toEqual(allInfo.LikeStorySubscriptionInputIdField);
    });
  });

  describe('directives', () => {
    it('core: name', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @include(if: true)
                #--------^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.IncludeDirective);
    });

    it('core: args', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @include(if: true)
                      #-------^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.IncludeDirectiveIfArg);
    });

    it('user defined: name', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @customDirective(if: true)
                #-------^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.CustomDirective);
    });

    it('user defined: args', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @customDirective(if: true)
                              #------^
            }
          \`
        `),
      });
      expect(info).toEqual(allInfo.CustomDirectiveIfArg);
    });

    it('unknown: name', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @xyzeagder(if: true)
                #-------^
            }
          \`
        `),
      });
      expect(info).toEqual(null);
    });

    it('unknown directive: args', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @xyzeagder(if: true)
                        #-------^
            }
          \`
        `),
      });
      expect(info).toEqual(null);
    });

    it('known directive: unknown args', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image @include(ifs: true)
                      #-------^
            }
          \`
        `),
      });
      expect(info).toEqual(null);
    });
  });

  describe('fragmentSpread', () => {
    it('fragment defined in same document', async () => {
      const info = await getInfo({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          const a = graphql\`
            fragment test on Player {
                ...Profile
            #-------^
            }

            fragment Profile on Player {
              id
              name
            }
          \`
        `),
      });
      expect(info).toMatchSnapshot();
    });

    it('fragment defined in same file', async () => {
      const info = await getInfo({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          const test = graphql\`
            fragment Profile on Player {
              id
              name
            }
          \`

          const a = graphql\`
            fragment test on Player {
              id
              ...Profile
          #--------^
            }
          \`
        `),
      });
      expect(info).toMatchSnapshot();
    });

    it('fragment defined in other file', async () => {
      const info = await getInfo({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          const a = graphql\`
            fragment test on Player {
              ...Profile
            #-----^
            }
          \`
        `),
        otherFiles: {
          'relay-modern/test2.js': `
            const b = graphql\`
              fragment Profile on Player {
                id
              }
            \`
          `,
        },
      });
      expect(info).toMatchSnapshot();
    });
  });

  describe('meta field', () => {
    describe('__typename', () => {
      it('when inside Object Type', async () => {
        const info = await getInfo({
          sourcePath: 'relay/test.js',
          ...code(`
            const a = Relay.QL\`
              fragment on Viewer {
                __typename
                #----^
              }
            \`
          `),
        });
        expect(info).toEqual(allInfo.TypeNameMeta);
      });

      it('when inside Interface Type', async () => {
        const info = await getInfo({
          sourcePath: 'relay/test.js',
          ...code(`
            const a = Relay.QL\`
              fragment on Node {
                __typename
                #----^
              }
            \`
          `),
        });
        expect(info).toEqual(allInfo.TypeNameMeta);
      });

      it('when inside Union Type', async () => {
        const info = await getInfo({
          sourcePath: 'relay/test.js',
          ...code(`
            const a = Relay.QL\`
              fragment on Entity {
                __typename
                #----^
              }
            \`
          `),
        });
        expect(info).toEqual(allInfo.TypeNameMeta);
      });
    });

    test('__schema', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            query Viewer {
              __schema
              #----^
            }
          \`
        `),
      });
      expect(info).toMatchSnapshot();
    });

    test('__type', async () => {
      const info = await getInfo({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            query Viewer {
              __type
              #--^
            }
          \`
        `),
      });
      expect(info).toMatchSnapshot();
    });
  });
});

async function getInfo(opts: {
  sourceText: string,
  sourcePath: string,
  position: GQLPosition,
  otherFiles?: { [name: string]: string },
}) {
  const rootDir = createTempFiles({
    '.gqlconfig': `
        {
          schema: {
            files: ['schema/*.gql', 'schema/*.graphql'],
            graphQLOptions: {
              commentDescriptions: true
            },
          },
          query: {
            files: [
              {
                match: ['query/**/*.graphql', 'query/**/*.gql'],
              },
              {
                match: 'relay/**/*.js',
                presets: ['relay'],
              },
              {
                match: 'relay-modern/**/*.js',
                presets: ['relay-modern'],
              },
              {
                match: 'apollo/**/*.js',
                presets: ['apollo'],
              },
            ]
          }
        }
      `,
    ...getSchemaFiles(),
    [opts.sourcePath]: opts.sourceText,
    ...opts.otherFiles,
  });

  const gql = new GQLService({
    configDir: rootDir,
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });

  await gql.start();

  const result = gql.getInfo({
    sourcePath: path.join(rootDir, opts.sourcePath),
    sourceText: opts.sourceText,
    position: opts.position,
  });

  await gql.stop();

  return result;
}
