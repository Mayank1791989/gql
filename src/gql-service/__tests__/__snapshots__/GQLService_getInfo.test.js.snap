// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query fragmentSpread fragment defined in other file 1`] = `
Object {
  "contents": Array [
    "fragment Profile on Player {
                id
              }",
  ],
}
`;

exports[`Query fragmentSpread fragment defined in same document 1`] = `
Object {
  "contents": Array [
    "fragment Profile on Player ",
  ],
}
`;

exports[`Query fragmentSpread fragment defined in same file 1`] = `
Object {
  "contents": Array [
    "fragment Profile on Player ",
  ],
}
`;

exports[`Query meta field __schema 1`] = `
Object {
  "contents": Array [
    "# Access the current type schema of this server.
  __schema: __Schema!",
    "# A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all
# available types and directives on the server, as well as the entry points for
# query, mutation, and subscription operations.
type __Schema {
  # A list of all types supported by this server.
  types: [__Type!]!

  # The type that query operations will be rooted at.
  queryType: __Type!

  # If this server supports mutation, the type that mutation operations will be rooted at.
  mutationType: __Type

  # If this server support subscription, the type that subscription operations will be rooted at.
  subscriptionType: __Type

  # A list of all directives supported by this server.
  directives: [__Directive!]!
}",
  ],
}
`;

exports[`Query meta field __type 1`] = `
Object {
  "contents": Array [
    "# Request the type information of a single type.
  __type(name: String!): __Type",
    "# The fundamental unit of any GraphQL Schema is the type. There are many kinds of
# types in GraphQL as represented by the \`__TypeKind\` enum.
#
# Depending on the kind of a type, certain fields describe information about that
# type. Scalar types provide no information beyond a name and description, while
# Enum types provide their values. Object and Interface types provide the fields
# they describe. Abstract types, Union and Interface, provide the Object types
# possible at runtime. List and NonNull types compose other types.
type __Type {
  kind: __TypeKind!
  name: String
  description: String
  fields(includeDeprecated: Boolean = false): [__Field!]
  interfaces: [__Type!]
  possibleTypes: [__Type!]
  enumValues(includeDeprecated: Boolean = false): [__EnumValue!]
  inputFields: [__InputValue!]
  ofType: __Type
}",
  ],
}
`;

exports[`Schema: getDef works in schema files 1`] = `
Object {
  "contents": Array [
    "type Viewer {
          name: string
        }",
  ],
}
`;
